smolyay/surrogate.py

```diff
@@ -1,546 +1,717 @@
-    (:math:y_0, :math:y_1, :math:y_2)
-    Then:
-    ..math::
-        \[
-        \begin{bmatrix}
-        B_0(x_0)& B_1(x_0) & B_2(x_0) \\
-        B_0(x_1)& B_1(x_1) & B_2(x_1) \\
-        B_0(x_2)& B_1(x_2) & B_2(x_2)
-        \end{bmatrix}
-        \begin{bmatrix}
-        C_0\\C_1\\C_2
-        \end{bmatrix}
-        =\begin{bmatrix}
-        F(y_0)\\F(y_1)\\F(y_2)
-        \end{bmatrix}
-        \]
-
-    Coefficients (x) can be obtained by solving linear equations,
-    since we have a full rank system of matrixes. The user can
-    select between 'lower upper decomposition' ('lu'), and
-    'inverse' ('inv') method for calculating the surrogate's (S)
-    coefficients. 'lstsq' solves the above equations through
-    least linear method.
-    S = C_0B_0(x) + C_1B_1(x) + C_2B_2(x)
+    domain: numpy.ndarray
+        the domain of the surrgoate
     """
 
-    def __init__(self, domain, grid_generator):
-        self.domain = domain
-        self.grid_generator = grid_generator
-        self._coefficients = None
-        self._grid = None
-        self._points = None
-        self._data = None
+    def __init__(self, domain):
+        self._domain = None
+        self._valid_cache = False
 
-    @property
-    def coefficients(self):
-        """list: coefficients of surrogate."""
-        if self._coefficients is not None:
-            return self._coefficients.tolist()
-        else:
-            return None
-
-    @property
-    def data(self):
-        """list: data at sampling grid points."""
-        if self._data is not None:
-            return self._data.tolist()
-        else:
-            return None
-
-    @property
-    def dimension(self):
-        """Numbers of independent variables."""
-        return self._domain.shape[0]
+        self.domain = domain
 
     @property
     def domain(self):
-        """list: domain of the function to be approximated."""
-        if self.dimension > 1:
-            return self._domain.tolist()
-        else:
-            return self._domain[0].tolist()
+        """domain: numpy.ndarray the domain of the surrogate."""
+        return self._domain
 
     @domain.setter
     def domain(self, value):
-        value = numpy.array(value, ndmin=2)
-        if value.ndim != 2 or value.shape[1] != 2:
-            raise TypeError('Domain must be dim x 2')
-        self._domain = value
-        self._reset_grid()
+        domain = numpy.sort(numpy.array(value, ndmin=2), axis=1)
+        if not numpy.array_equal(self._domain, domain):
+            self._domain = domain
+            self._valid_cache = False
 
     @property
-    def grid(self):
-        """:class:`IndexGrid` Sampling grid."""
-        if self._grid is None:
-            self._grid = self.grid_generator(self.dimension)
-        return self._grid
+    def num_dimensions(self):
+        """int: number of independent variables."""
+        return self.domain.shape[0]
+    
+    def _in_domain(self, X):
```

Comment:
This functions like an assert since it doesn't return true / false, so I would name it to include that.
```suggestion
    def _assert_in_domain(self, X):
```